lib/native/source/libboolector/interface_wrap.c

```diff
@@ -3654,137 +3652,157 @@ SWIGEXPORT jobjectArray JNICALL Java_org_sosy_1lab_java_1smt_solvers_boolector_B
   }
 
   if (buffer) {
-    foutString = buffer;
+    buffer[length] = '\0';
+    fileOutString = buffer;
   } else {
-    foutString = "";
+    fileOutString = "";
   }
-  
+
   (*jenv)->SetObjectArrayElement(jenv, jniArray, 0, (*jenv)->NewStringUTF(jenv, resultString));
-  
-  (*jenv)->SetObjectArrayElement(jenv, jniArray, 1, (*jenv)->NewStringUTF(jenv, (const char *)foutString));
-  
+
+  (*jenv)->SetObjectArrayElement(jenv, jniArray, 1, (*jenv)->NewStringUTF(jenv, (const char *)fileOutString));
+
   (*jenv)->SetObjectArrayElement(jenv, jniArray, 2, (*jenv)->NewStringUTF(jenv, errormsg));
-  
+
   (*jenv)->SetObjectArrayElement(jenv, jniArray, 3, (*jenv)->NewStringUTF(jenv, statusString));
-  
+
   (*jenv)->SetObjectArrayElement(jenv, jniArray, 4, (*jenv)->NewStringUTF(jenv, flagString));
 
   (*jenv)->DeleteLocalRef(jenv, classString);
+  unlink(tempfileNameOut);
+  fclose(fileOut);
   free(buffer);
-  unlink(tfnwdIn);
-  free(tfnwdIn);
-  fclose(fparse);
-  unlink(tfnwdOut);  
-  free(tfnwdOut);
-  fclose(fout);
   free(statusString);
   free(resultString);
   return jniArray;
 }
-    
+
 //dumps NODE into new file and reads it to give it back
 SWIGEXPORT jstring JNICALL Java_org_sosy_1lab_java_1smt_solvers_boolector_BtorJNI_boolector_1help_1dump_1node_1smt2(JNIEnv *jenv, jclass jcls, jlong jarg1, jlong jarg2) {
   jstring jresult = 0;
   Btor *arg1 = (Btor *) 0 ;
-  char *result = 0 ;
-  char filename[] = "boolector_help_dump_node_smt2_tempinfile-XXXXXX";
-  FILE *f = 0;
-  char * buffer = 0;
-  int fd = -1;
-  long length = 0;
+  char filenameTemplate[] = "boolector_help_dump_node_smt2_tempinfile-XXXXXX";
+  FILE *file = 0;
+  char *buffer = NULL;
+  int fileDesrc = -1;
+  long fileLength = 0;
   BoolectorNode *arg2 = (BoolectorNode *) 0 ;
-  char *tfnwd = addTemppathToFilename(filename);
+  char *tempfileName = addTemppathToFilename(filenameTemplate);
 
-  fd = mkstemp(tfnwd);
-  if(fd == -1) {
-    free(tfnwd);
-    perror("ERROR CREATING TEMPORARY FILE FOR");
+  if (tempfileName == NULL) {
+    perror("ERROR CREATING TEMPORARY FILE FOR BOOLECTOR_HELP_DUMP_NODE_SMT2");
+    SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "FileName for boolector_help_dump_node_smt2 may not be NULL");
+    return 0;
+  }
+
+  fileDesrc = mkstemp(tempfileName);
+  if (fileDesrc == -1) {
+    free(tempfileName);
+    perror("ERROR CREATING TEMPORARY FILE FOR BOOLECTOR_HELP_DUMP_NODE_SMT2");
     SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "FileDescriptor for boolector_help_dump_node_smt2 may not be NULL");
     return 0;
   }
-    
+
   (void)jenv;
   (void)jcls;
   arg1 = *(Btor **)&jarg1;
-  arg2 = *(BoolectorNode **)&jarg2; 
+  arg2 = *(BoolectorNode **)&jarg2;
 
-  f = fdopen(fd, "w+");
-  if(f==NULL) {
-    free(tfnwd);
-    unlink(tfnwd);
-    perror("ERROR: COULDNT DUMP NODE BECAUSE IT COULDNT CREATE A DUMP FILE"); 
+  file = fdopen(fileDesrc, "w+");
+  unlink(tempfileName);
```

Comment:
Why do we need to unlink the file here?

Reply:
We don't, we could do it later, but it doesn't matter because we have the filepointer. The file only gets deleted after we close the opened file and Boolector does not close the file itself. I read that for safety reasons you should do this as soon as possible though.